libfwupd/fwupd-client.c

```diff
@@ -121,13 +124,127 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(FwupdCurlHelper, fwupd_client_curl_helper_free)
 #endif
 
+typedef struct {
+	FwupdClient	*self;
+	gchar		*property_name;
+	guint		 signal_id;
+	FwupdDevice	*device;
+} FwupdClientContextHelper;
+
+static void
+fwupd_client_context_helper_free (FwupdClientContextHelper *helper)
+{
+	g_clear_object (&helper->device);
+	g_object_unref (helper->self);
+	g_free (helper->property_name);
+	g_free (helper);
+}
+
+/* always executed in the main context given by priv->main_ctx */
+static void
+fwupd_client_context_object_notify (FwupdClient *self, const gchar *property_name)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+
+	g_return_if_fail (g_main_context_is_owner (priv->main_ctx));
+
+	/* property */
+	g_object_notify (G_OBJECT (self), property_name);
+
+	/* legacy signal name */
+	if (g_strcmp0 (property_name, "status") == 0)
+		g_signal_emit (self, signals[SIGNAL_STATUS_CHANGED], 0, priv->status);
+}
+
+/* emits all pending context helpers in the correct GMainContext */
+static gboolean
+fwupd_client_context_idle_cb (gpointer user_data)
+{
+	FwupdClient *self = FWUPD_CLIENT (user_data);
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_mutex);
+
+	for (guint i = 0; i < priv->idle_sources->len; i++) {
+		FwupdClientContextHelper *helper = g_ptr_array_index (priv->idle_sources, i);
+
+		/* property */
+		if (helper->property_name != NULL)
+			fwupd_client_context_object_notify (self, helper->property_name);
+
+		/* device signal */
+		if (helper->signal_id !=0 && helper->device != NULL)
+			g_signal_emit (self, signals[helper->signal_id], 0, helper->device);
+	}
+
+	/* all done */
+	g_ptr_array_set_size (priv->idle_sources, 0);
+	priv->idle_id = 0;
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fwupd_client_context_helper (FwupdClient *self, FwupdClientContextHelper *helper)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_mutex);
+	g_autoptr(GSource) source = g_idle_source_new ();
+
+	/* run in the correct GMainContext and thread */
+	g_source_set_callback (source, fwupd_client_context_idle_cb, self, NULL);
+	if (priv->idle_id != 0)
+		g_source_remove (priv->idle_id);
```

Comment:
If there’s already a source attached to the context (i.e. `priv->idle_id != 0`) then you could just return early from this function, don’t remove the existing source, and don’t add a new one. Would save a bit of locking churn on that `GMainContext`

Reply:
Agree. Something like this?

	FwupdClientPrivate *priv = GET_PRIVATE (self);
	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_mutex);

	/* already a source attached to the context */
	if (priv->idle_id == 0) {
		g_autoptr(GSource) source = g_idle_source_new ();
		g_source_set_callback (source, fwupd_client_context_idle_cb, self, NULL);
		priv->idle_id = g_source_attach (g_steal_pointer (&source), priv->main_ctx);
	}

	/* run in the correct GMainContext and thread */
	g_ptr_array_add (priv->idle_sources, helper);

Reply:
Yes, although I think the `already a source attached to the context` comment needs inverting (a source isn’t attached if that branch is taken)

Reply:
Ohh yes. indeed. thanks!